File: impl/sdk/rs/resonate/src/transport.rs
use std::sync::Arc;

use crate::error::{Error, Result};
use crate::network::Network;

/// The Transport wraps a Network to add JSON serialization, deserialization,
/// and correlation validation. Resonate and its sub-components use the
/// transport—never the raw network.
#[derive(Clone)]
pub struct Transport {
    network: Arc<dyn Network>,
}

/// A parsed incoming message from the network.
///
/// Mirrors the TS types:
/// ```ts
/// type ExecuteMsg = { kind: "execute"; data: { task: { id: string; version: number } } };
/// type UnblockMsg = { kind: "unblock"; data: { promise: PromiseRecord } };
/// ```
#[derive(Debug, Clone, serde::Deserialize)]
#[serde(tag = "kind")]
pub enum Message {
    #[serde(rename = "execute")]
    Execute(ExecuteMsg),
    #[serde(rename = "unblock")]
    Unblock(UnblockMsg),
}

/// Execute message — server tells this worker to run a task.
///
/// JSON shape: `{ kind: "execute", data: { task: { id, version } } }`
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ExecuteMsg {
    pub data: ExecuteData,
}

impl ExecuteMsg {
    /// Task ID — shorthand for `data.task.id`.
    pub fn task_id(&self) -> &str {
        &self.data.task.id
    }
    /// Task version — shorthand for `data.task.version`.
    pub fn version(&self) -> i64 {
        self.data.task.version
    }
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ExecuteData {
    pub task: TaskRef,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct TaskRef {
    pub id: String,
    #[serde(default)]
    pub version: i64,
}

/// Unblock message — a promise this worker is waiting on has been settled.
///
/// JSON shape: `{ kind: "unblock", data: { promise: PromiseRecord } }`
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct UnblockMsg {
    pub data: UnblockData,
}

impl UnblockMsg {
    /// The settled promise — shorthand for `data.promise`.
    pub fn promise(&self) -> &serde_json::Value {
        &self.data.promise
    }
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct UnblockData {
    pub promise: serde_json::Value,
}

/// Extract the `data` portion from a protocol envelope response.
pub fn response_data(resp: &serde_json::Value) -> Result<&serde_json::Value> {
    resp.get("data")
        .ok_or_else(|| Error::DecodingError("response missing 'data' envelope field".into()))
}

/// Extract the `head.status` from a protocol envelope response.
pub fn response_status(resp: &serde_json::Value) -> Result<u64> {
    resp.get("head")
        .and_then(|h| h.get("status"))
        .and_then(|s| s.as_u64())
        .ok_or_else(|| Error::DecodingError("response missing 'head.status' envelope field".into()))
}

impl Transport {
    /// Build a Transport from a Network.
    pub fn new(network: Arc<dyn Network>) -> Self {
        Self { network }
    }

    /// Send an already-serialized request through the network, returning the parsed response.
    /// Validates that `response.kind == kind` and `response.head.corrId == corr_id`.
    pub async fn send(
        &self,
        kind: &str,
        corr_id: &str,
        body: &str,
    ) -> Result<serde_json::Value> {
        tracing::debug!(direction = "send_req", body = %body, "transport");

        let resp_str = self.network.send(body.to_owned()).await?;
        tracing::debug!(direction = "send_res", body = %resp_str, "transport");

        let response: serde_json::Value = serde_json::from_str(&resp_str).map_err(|e| {
            Error::DecodingError(format!("invalid response JSON: {e}, resp: {resp_str}"))
        })?;

        // Validate kind matches
        let resp_kind = response
            .get("kind")
            .and_then(|k| k.as_str())
            .unwrap_or("");
        if resp_kind != kind {
            return Err(Error::ServerError {
                code: 500,
                message: format!(
                    "response kind mismatch: expected '{}', got '{}'",
                    kind, resp_kind
                ),
            });
        }

        // Validate corrId matches
        let resp_corr = response
            .get("head")
            .and_then(|h| h.get("corrId"))
            .and_then(|c| c.as_str())
            .unwrap_or("");
        if resp_corr != corr_id {
            return Err(Error::ServerError {
                code: 500,
                message: format!(
                    "response corrId mismatch: expected '{}', got '{}'",
                    corr_id, resp_corr
                ),
            });
        }

        Ok(response)
    }

    /// Convenience: serialize a `serde_json::Value` envelope and send it.
    /// Extracts `kind` and `head.corrId` from the value before delegating to [`send`].
    pub async fn send_json(&self, request: serde_json::Value) -> Result<serde_json::Value> {
        let kind = request
            .get("kind")
            .and_then(|k| k.as_str())
            .unwrap_or("");
        let corr_id = request
            .get("head")
            .and_then(|h| h.get("corrId"))
            .and_then(|c| c.as_str())
            .unwrap_or("");
        let body = serde_json::to_string(&request)?;
        self.send(kind, corr_id, &body).await
    }

    /// Register a callback for incoming messages. Parses JSON → Message,
    /// discards invalid messages, forwards valid ones.
    pub fn recv(&self, callback: Box<dyn Fn(Message) + Send + Sync>) {
        self.network.recv(Box::new(move |raw: String| {
            match serde_json::from_str::<Message>(&raw) {
                Ok(msg) => {
                    tracing::debug!(direction = "recv", body = %raw, "transport");
                    callback(msg)
                }
                Err(e) => {
                    tracing::warn!(error = %e, raw = %raw, "failed to parse incoming message");
                }
            }
        }));
    }

    /// Access the underlying network.
    pub fn network(&self) -> &Arc<dyn Network> {
        &self.network
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::network::LocalNetwork;

    #[tokio::test]
    async fn transport_send_and_validate_envelope_format() {
        let net = Arc::new(LocalNetwork::new(Some("test".into()), None));
        let transport = Transport::new(net);

        // Envelope format request — pre-serialized
        let body = serde_json::json!({
            "kind": "promise.create",
            "head": {
                "corrId": "env123",
                "version": "2025-01-15",
            },
            "data": {
                "id": "p2",
                "timeoutAt": i64::MAX,
                "param": {},
                "tags": {},
            },
        });
        let body_str = serde_json::to_string(&body).unwrap();

        let resp = transport
            .send("promise.create", "env123", &body_str)
            .await
            .unwrap();
        assert_eq!(resp["kind"], "promise.create");
        assert_eq!(resp["head"]["corrId"], "env123");
        assert_eq!(resp["data"]["promise"]["id"], "p2");
    }
}

File: impl/sdk/py/src/resonate/transport.py
from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

import msgspec

from resonate.error import DecodingError, ServerError
from resonate.observability import Dropped, logging_observer
from resonate.types import PromiseRecord

if TYPE_CHECKING:
    from collections.abc import Callable, Sequence

    from resonate_base.connections import Network, Source

    from resonate.observability import Observer

logger = logging.getLogger(__name__)


# =============================================================================
# Incoming messages (recv path)
# =============================================================================


class TaskRef(msgspec.Struct, kw_only=True, frozen=True):
    id: str
    version: int = msgspec.field(default=0)


class ExecuteData(msgspec.Struct, kw_only=True, frozen=True):
    task: TaskRef


class ExecuteMsg(
    msgspec.Struct, tag="execute", tag_field="kind", kw_only=True, frozen=True
):
    data: ExecuteData

    @property
    def task_id(self) -> str:
        return self.data.task.id

    @property
    def version(self) -> int:
        return self.data.task.version


class UnblockData(msgspec.Struct, kw_only=True, frozen=True):
    promise: PromiseRecord


class UnblockMsg(
    msgspec.Struct, tag="unblock", tag_field="kind", kw_only=True, frozen=True
):
    data: UnblockData

    @property
    def promise(self) -> PromiseRecord:
        """Return the settled promise -- shorthand for ``data.promise``."""
        return self.data.promise


# A parsed incoming message from the network, discriminated by its ``kind``
# field.
Message = ExecuteMsg | UnblockMsg


# =============================================================================
# Outgoing responses (send path)
# =============================================================================


class ResponseHead(msgspec.Struct, kw_only=True, frozen=True, rename="camel"):
    """The ``head`` of a protocol response envelope.

    Every field is defaulted: a server that omits ``head`` entirely, or any
    field within it, is a *valid* response meaning "200, no correlation echo",
    and the mismatch checks in :meth:`Transport.send` catch the cases that
    matter. Defaulting here is what lets the rest of the SDK read
    ``resp.head.status`` unconditionally instead of re-deriving it behind
    ``isinstance`` guards at every call site.
    """

    corr_id: str = ""
    status: int = 200


class Response(msgspec.Struct, kw_only=True, frozen=True):
    """A parsed protocol response envelope: ``{ kind, head, data }``.

    The single parse boundary for everything arriving over the
    :class:`~resonate_base.connections.Network`. Below this point the SDK works
    with a typed value and never asks "is this a dict?" again -- the shape
    questions are all answered here, once.
    """

    kind: str = ""
    head: ResponseHead = msgspec.field(default_factory=ResponseHead)
    data: Any = msgspec.field(default_factory=dict)


# =============================================================================
# Transport
# =============================================================================


class Transport:
    """Adds JSON serialization, deserialization, and correlation validation.

    Resonate and its sub-components use the transport -- never the raw
    connections. Requests go out over the single ``network``; push messages
    come in over every ``source``.

    ``observer`` receives a :class:`~resonate.observability.Dropped` event for
    each incoming push message that cannot be parsed, so the drop is
    assertable rather than merely logged.
    """

    def __init__(
        self,
        network: Network,
        sources: Sequence[Source] = (),
        observer: Observer = logging_observer,
    ) -> None:
        self._network = network
        self._sources = tuple(sources)
        self._observer = observer

    async def send(
        self,
        kind: str,
        corr_id: str,
        body: str,
        headers: dict[str, str] | None = None,
    ) -> Response:
        """Send an already-serialized request, returning the parsed response.

        Parses at the edge: the raw JSON is decoded straight into
        :class:`Response`, so callers receive a typed value whose ``status``
        and ``corrId`` have already been validated.

        ``headers`` carry the request metadata, including the routing origin
        under :data:`~resonate_base.ORIGIN_HEADER`, and are passed straight to
        the connection (see :meth:`~resonate_base.connections.Network.send`) so
        a sharding substrate never has to open the payload. The caller owns
        that folding -- the transport only moves what it is handed.
        """
        logger.debug("transport send_req: %s", body)

        resp_str = await self._network.send(body, headers)
        logger.debug("transport send_res: %s", resp_str)

        try:
            response = msgspec.json.decode(resp_str, type=Response)
        except msgspec.MsgspecError as exc:
            msg = f"invalid response JSON: {exc}, resp: {resp_str}"
            raise DecodingError(msg) from exc

        if response.kind != kind:
            msg = f"response kind mismatch: expected '{kind}', got '{response.kind}'"
            raise ServerError(500, msg)

        if response.head.corr_id != corr_id:
            msg = (
                f"response corrId mismatch: expected '{corr_id}', "
                f"got '{response.head.corr_id}'"
            )
            raise ServerError(500, msg)

        return response

    def recv(self, callback: Callable[[Message], None]) -> None:
        """Register a callback for incoming messages on every source."""

        def on_raw(raw: str) -> None:
            try:
                msg = msgspec.json.decode(raw, type=Message)
            except msgspec.MsgspecError as exc:
                self._observer(Dropped(what="incoming-message", id="", cause=str(exc)))
                logger.warning(
                    "failed to parse incoming message: %s; raw: %s", exc, raw
                )
                return
            logger.debug("transport recv: %s", raw)
            callback(msg)

        for source in self._sources:
            source.recv(on_raw)

    def network(self) -> Network:
        """Access the underlying network."""
        return self._network
